File: library/src/main/java/com/tundem/aboutlibraries/Libs.java
package com.tundem.aboutlibraries;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.text.TextUtils;
import android.util.Log;

import com.tundem.aboutlibraries.detector.Detect;
import com.tundem.aboutlibraries.entity.Library;
import com.tundem.aboutlibraries.entity.License;
import com.tundem.aboutlibraries.util.Util;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Libs {
    public static final String BUNDLE_FIELDS = "ABOUT_LIBRARIES_FIELDS";
    public static final String BUNDLE_LIBS = "ABOUT_LIBRARIES_LIBS";
    public static final String BUNDLE_EXCLUDE_LIBS = "ABOUT_LIBRARIES_EXCLUDE_LIBS";
    public static final String BUNDLE_AUTODETECT = "ABOUT_LIBRARIES_AUTODETECT";
    public static final String BUNDLE_SORT = "ABOUT_LIBRARIES_SORT";

    public static final String BUNDLE_LICENSE = "ABOUT_LIBRARIES_LICENSE";
    public static final String BUNDLE_LICENSE_DIALOG = "ABOUT_LIBRARIES_LICENSE_DIALOG";
    public static final String BUNDLE_VERSION = "ABOUT_LIBRARIES_VERSION";

    public static final String BUNDLE_THEME = "ABOUT_LIBRARIES_THEME";
    public static final String BUNDLE_TITLE = "ABOUT_LIBRARIES_TITLE";
    public static final String BUNDLE_ACCENT_COLOR = "ABOUT_LIBRARIES_ACCENT_COLOR";
    public static final String BUNDLE_TRANSLUCENT_DECOR = "ABOUT_LIBRARIES_TRANSLUCENT_DECOR";

    private static final String DEFINE_LICENSE = "define_license_";
    private static final String DEFINE_INT = "define_int_";
    private static final String DEFINE_EXT = "define_";

    private static Context ctx;
    private static Libs libs = null;

    private ArrayList<Library> internLibraries = new ArrayList<Library>();
    private ArrayList<Library> externLibraries = new ArrayList<Library>();
    private ArrayList<License> licenses = new ArrayList<License>();

    private Libs() {
        String[] fields = toStringArray(R.string.class.getFields());
        init(fields);
    }

    private Libs(String[] fields) {
        init(fields);
    }

    /**
     * init method
     *
     * @param fields
     */
    private void init(String[] fields) {
        ArrayList<String> foundLicenseIdentifiers = new ArrayList<String>();
        ArrayList<String> foundInternalLibraryIdentifiers = new ArrayList<String>();
        ArrayList<String> foundExternalLibraryIdentifiers = new ArrayList<String>();

        if (fields != null) {
            for (int i = 0; i < fields.length; i++) {
                if (fields[i].startsWith(DEFINE_LICENSE)) {
                    foundLicenseIdentifiers.add(fields[i].replace(DEFINE_LICENSE, ""));
                } else if (fields[i].startsWith(DEFINE_INT)) {
                    foundInternalLibraryIdentifiers.add(fields[i].replace(DEFINE_INT, ""));
                } else if (fields[i].startsWith(DEFINE_EXT)) {
                    foundExternalLibraryIdentifiers.add(fields[i].replace(DEFINE_EXT, ""));
                }
            }
        }

        //add licenses
        for (String licenseIdentifier : foundLicenseIdentifiers) {
            License license = genLicense(licenseIdentifier);
            if (license != null) {
                licenses.add(license);
            }
        }
        //add internal libs
        for (String internalIdentifier : foundInternalLibraryIdentifiers) {
            Library library = genLibrary(internalIdentifier);
            if (library != null) {
                library.setInternal(true);
                internLibraries.add(library);
            }
        }

        //add external libs
        for (String externalIdentifier : foundExternalLibraryIdentifiers) {
            Library library = genLibrary(externalIdentifier);
            if (library != null) {
                library.setInternal(false);
                externLibraries.add(library);
            }
        }
    }


    /**
     * A helper method to get a String[] out of a fieldArray
     *
     * @param fields R.strings.class.getFields()
     * @return a String[] with the string ids we need
     */
    public static String[] toStringArray(Field[] fields) {
        ArrayList<String> fieldArray = new ArrayList<String>();
        for (Field field : fields) {
            if (field.getName().contains(DEFINE_EXT)) {
                fieldArray.add(field.getName());
            }
        }
        return fieldArray.toArray(new String[fieldArray.size()]);
    }

    /**
     * Creates a new Instance of Libs. This only includes internal internLibraries
     *
     * @param context
     * @return
     */
    public static Libs getInstance(Context context) {
        ctx = context;
        if (libs == null) {
            libs = new Libs();
        }
        return libs;
    }

    /**
     * Creates a new Instance of Libs. This give fields[] also contains the ressources of any lib you've used
     * Retrieve the fields[] by calling following method: Field[] fields = R.string.class.getFields();
     *
     * @param context
     * @param fields
     * @return
     */
    public static Libs getInstance(Context context, String[] fields) {
        ctx = context;
        if (libs == null) {
            libs = new Libs(fields);
        }
        return libs;
    }


    /**
     * This will summarize all libraries and elimate duplicates
     *
     * @param internalLibraries the String[] with the internalLibraries (if set manual)
     * @param excludeLibraries  the String[] with the libs to be excluded
     * @param autoDetect        defines if the libraries should be resolved by their classpath (if possible)
     * @param sort              defines if the array should be sorted
     * @return the summarized list of included Libraries
     */
    public ArrayList<Library> prepareLibraries(String[] internalLibraries, String[] excludeLibraries, boolean autoDetect, boolean sort) {
        HashMap<String, Library> libraries = new HashMap<String, Library>();

        if (autoDetect) {
            for (Library lib : libs.getAutoDetectedLibraries()) {
                libraries.put(lib.getDefinedName(), lib);
            }
        }

        //Add all external libraries
        for (Library lib : libs.getExternLibraries()) {
            libraries.put(lib.getDefinedName(), lib);
        }

        //Now add all libs which do not contains the info file, but are in the AboutLibraries lib
        if (internalLibraries != null) {
            for (String internalLibrary : internalLibraries) {
                Library lib = libs.getLibrary(internalLibrary);
                if (lib != null) {
                    libraries.put(lib.getDefinedName(), lib);
                }
            }
        }

        ArrayList<Library> resultLibraries = new ArrayList<Library>(libraries.values());

        //remove libraries which should be excluded
        if (excludeLibraries != null) {
            List<Library> libsToRemove = new ArrayList<Library>();
            for (String excludeLibrary : excludeLibraries) {
                for (Library library : resultLibraries) {
                    if (library.getDefinedName().equals(excludeLibrary)) {
                        libsToRemove.add(library);
                        break;
                    }
                }
            }
            for (Library libToRemove : libsToRemove) {
                resultLibraries.remove(libToRemove);
            }
        }

        if (sort) {
            Collections.sort(resultLibraries);
        }
        return resultLibraries;
    }

    /**
     * Get all autoDetected Libraries
     *
     * @return an ArrayList<Library> with all found libs by their classpath
     */
    public ArrayList<Library> getAutoDetectedLibraries() {
        ArrayList<Library> libraries = new ArrayList<Library>();

        PackageInfo pi = Util.getPackageInfo(ctx);
        if (pi != null) {
            String[] autoDetectedLibraries = ctx.getSharedPreferences("aboutLibraries_" + pi.versionCode, Context.MODE_PRIVATE).getString("autoDetectedLibraries", "").split(";");

            if (autoDetectedLibraries.length > 0) {
                for (String autoDetectedLibrary : autoDetectedLibraries) {
                    Library lib = getLibrary(autoDetectedLibrary);
                    if (lib != null) {
                        libraries.add(lib);
                    }
                }
            }
        }

        if (libraries.size() == 0) {
            String delimiter = "";
            String autoDetectedLibrariesPref = "";
            for (Library lib : Detect.detect(ctx, getLibraries())) {
                libraries.add(lib);

                autoDetectedLibrariesPref = autoDetectedLibrariesPref + delimiter + lib.getDefinedName();
                delimiter = ";";
            }

            if (pi != null) {
                ctx.getSharedPreferences("aboutLibraries_" + pi.versionCode, Context.MODE_PRIVATE).edit().putString("autoDetectedLibraries", autoDetectedLibrariesPref).commit();
            }
        }

        return libraries;
    }

    /**
     * Get all intern available Libraries
     *
     * @return an ArrayList<Library> with all available internLibraries
     */
    public ArrayList<Library> getInternLibraries() {
        return new ArrayList<Library>(internLibraries);
    }

    /**
     * Get all extern available Libraries
     *
     * @return an ArrayList<Library> with all available externLibraries
     */
    public ArrayList<Library> getExternLibraries() {
        return new ArrayList<Library>(externLibraries);
    }

    /**
     * Get all available licenses
     *
     * @return an ArrayLIst<License> with all available Licenses
     */
    public ArrayList<License> getLicenses() {
        return new ArrayList<License>(licenses);
    }

    /**
     * Get all available Libraries
     *
     * @return an ArrayList<Library> with all available Libraries
     */
    public ArrayList<Library> getLibraries() {
        ArrayList<Library> libs = new ArrayList<Library>();
        libs.addAll(getInternLibraries());
        libs.addAll(getExternLibraries());
        return libs;
    }

    /**
     * Get a library by its name (the name must be equal)
     *
     * @param libraryName the name of the lib (NOT case sensitiv) or the real name of the lib (this is the name used for github)
     * @return the found library or null
     */
    public Library getLibrary(String libraryName) {
        for (Library library : getLibraries()) {
            if (library.getLibraryName().toLowerCase().equals(libraryName.toLowerCase())) {
                return library;
            } else if (library.getDefinedName().toLowerCase().equals(libraryName.toLowerCase())) {
                return library;
            }
        }
        return null;
    }

    /**
     * Find a library by a searchTerm (Limit the results if there are more than one)
     *
     * @param searchTerm the term which is in the libs name (NOT case sensitiv) or the real name of the lib (this is the name used for github)
     * @param limit      -1 for all results or > 0 for a limitted result
     * @return an ArrayList<Library> with the found internLibraries
     */
    public ArrayList<Library> findLibrary(String searchTerm, int limit) {
        ArrayList<Library> localLibs = new ArrayList<Library>();

        int count = 0;
        for (Library library : getLibraries()) {
            if (library.getLibraryName().toLowerCase().contains(searchTerm.toLowerCase()) || library.getDefinedName().toLowerCase().contains(searchTerm.toLowerCase())) {
                localLibs.add(library);
                count = count + 1;

                if (limit != -1 && limit < count) {
                    break;
                }
            }
        }

        return localLibs;
    }


    public License getLicense(String licenseName) {
        for (License license : getLicenses()) {
            if (license.getLicenseName().toLowerCase().equals(licenseName.toLowerCase())) {
                return license;
            } else if (license.getDefinedName().toLowerCase().equals(licenseName.toLowerCase())) {
                return license;
            }
        }
        return null;
    }

    private License genLicense(String licenseName) {
        licenseName = licenseName.replace("-", "_");

        try {
            License lic = new License();
            lic.setDefinedName(licenseName);
            lic.setLicenseName(getStringResourceByName("license_" + licenseName + "_licenseName"));
            lic.setLicenseWebsite(getStringResourceByName("license_" + licenseName + "_licenseWebsite"));
            lic.setLicenseShortDescription(getStringResourceByName("license_" + licenseName + "_licenseShortDescription"));
            lic.setLicenseDescription(getStringResourceByName("license_" + licenseName + "_licenseDescription"));
            return lic;
        } catch (Exception ex) {
            Log.e("com.tundem.aboutlibraries", "Failed to generateLicense from file: " + ex.toString());
            return null;
        }
    }

    private Library genLibrary(String libraryName) {
        libraryName = libraryName.replace("-", "_");

        try {
            Library lib = new Library();

            //Get custom vars to insert into defined areas
            HashMap<String, String> customVariables = getCustomVariables(libraryName);

            lib.setDefinedName(libraryName);
            lib.setAuthor(getStringResourceByName("library_" + libraryName + "_author"));
            lib.setAuthorWebsite(getStringResourceByName("library_" + libraryName + "_authorWebsite"));
            lib.setLibraryName(getStringResourceByName("library_" + libraryName + "_libraryName"));
            lib.setLibraryDescription(insertVariables(getStringResourceByName("library_" + libraryName + "_libraryDescription"), customVariables));
            lib.setLibraryVersion(getStringResourceByName("library_" + libraryName + "_libraryVersion"));
            lib.setLibraryWebsite(getStringResourceByName("library_" + libraryName + "_libraryWebsite"));

            String licenseId = getStringResourceByName("library_" + libraryName + "_licenseId");
            if (TextUtils.isEmpty(licenseId)) {
                License license = new License();
                license.setLicenseName(getStringResourceByName("library_" + libraryName + "_licenseVersion"));
                license.setLicenseWebsite(getStringResourceByName("library_" + libraryName + "_licenseLink"));
                license.setLicenseShortDescription(insertVariables(getStringResourceByName("library_" + libraryName + "_licenseContent"), customVariables));
                lib.setLicense(license);
            } else {
                License license = getLicense(licenseId);
                if (license != null) {
                    license = license.copy();
                    license.setLicenseShortDescription(insertVariables(license.getLicenseShortDescription(), customVariables));
                    license.setLicenseDescription(insertVariables(license.getLicenseDescription(), customVariables));
                    lib.setLicense(license);
                }
            }

            lib.setOpenSource(Boolean.valueOf(getStringResourceByName("library_" + libraryName + "_isOpenSource")));
            lib.setRepositoryLink(getStringResourceByName("library_" + libraryName + "_repositoryLink"));

            lib.setClassPath(getStringResourceByName("library_" + libraryName + "_classPath"));

            if (TextUtils.isEmpty(lib.getLibraryName()) && TextUtils.isEmpty(lib.getLibraryDescription())) {
                return null;
            }

            return lib;
        } catch (Exception ex) {
            Log.e("com.tundem.aboutlibraries", "Failed to generateLibrary from file: " + ex.toString());
            return null;
        }
    }

    public HashMap<String, String> getCustomVariables(String libraryName) {
        HashMap<String, String> customVariables = new HashMap<String, String>();

        String customVariablesString = getStringResourceByName(DEFINE_EXT + libraryName);
        if (TextUtils.isEmpty(customVariablesString)) {
            customVariablesString = getStringResourceByName(DEFINE_INT + libraryName);
        }

        if (!TextUtils.isEmpty(customVariablesString)) {
            String[] customVariableArray = customVariablesString.split(";");
            if (customVariableArray.length > 0) {
                for (String customVariableKey : customVariableArray) {
                    String customVariableContent = getStringResourceByName("library_" + libraryName + "_" + customVariableKey);
                    if (!TextUtils.isEmpty(customVariableContent)) {
                        customVariables.put(customVariableKey, customVariableContent);
                    }
                }
            }
        }

        return customVariables;
    }

    public String insertVariables(String insertInto, HashMap<String, String> variables) {
        for (Map.Entry<String, String> entry : variables.entrySet()) {
            insertInto = insertInto.replace("<<<" + entry.getKey().toUpperCase() + ">>>", entry.getValue());
        }
        return insertInto;
    }

    public String getStringResourceByName(String aString) {
        String packageName = ctx.getPackageName();

        int resId = ctx.getResources().getIdentifier(aString, "string", packageName);
        if (resId == 0) {
            return "";
        } else {
            return ctx.getString(resId);
        }
    }
}

File: library/src/main/java/com/mikepenz/aboutlibraries/Libs.kt
@file:JvmName("Libs")

package com.mikepenz.aboutlibraries

import android.content.Context
import android.text.TextUtils
import android.util.Log
import com.mikepenz.aboutlibraries.detector.Detect
import com.mikepenz.aboutlibraries.entity.Library
import com.mikepenz.aboutlibraries.entity.License
import com.mikepenz.aboutlibraries.util.getFields
import com.mikepenz.aboutlibraries.util.getPackageInfo
import java.util.*
import kotlin.collections.ArrayList

public class Libs {

    private val internLibraries = ArrayList<Library>()
    private val externLibraries = ArrayList<Library>()
    private val licenses = ArrayList<License>()

    /**
     * Get all available Libraries
     *
     * @return an ArrayList Library with all available Libraries
     */
    val libraries: ArrayList<Library>
        get() {
            val libs = ArrayList<Library>()
            libs.addAll(getInternLibraries())
            libs.addAll(getExternLibraries())
            return libs
        }

    enum class LibraryFields {
        AUTHOR_NAME,
        AUTHOR_WEBSITE,
        LIBRARY_NAME,
        LIBRARY_DESCRIPTION,
        LIBRARY_VERSION,
        LIBRARY_WEBSITE,
        LIBRARY_OPEN_SOURCE,
        LIBRARY_REPOSITORY_LINK,
        LIBRARY_CLASSPATH,
        LICENSE_NAME,
        LICENSE_SHORT_DESCRIPTION,
        LICENSE_DESCRIPTION,
        LICENSE_WEBSITE
    }

    enum class ActivityStyle {
        LIGHT,
        DARK,
        LIGHT_DARK_TOOLBAR
    }

    enum class SpecialButton {
        SPECIAL1,
        SPECIAL2,
        SPECIAL3
    }

    constructor(context: Context) {
        init(context, context.getFields())
    }

    constructor(context: Context, fields: Array<String>?) {
        init(context, fields)
    }

    /**
     * init method
     *
     * @param fields
     */
    private fun init(ctx: Context, fields: Array<String>?) {
        val foundLicenseIdentifiers = ArrayList<String>()
        val foundInternalLibraryIdentifiers = ArrayList<String>()
        val foundExternalLibraryIdentifiers = ArrayList<String>()

        if (fields != null) {
            for (field in fields) {
                if (field.startsWith(DEFINE_LICENSE)) {
                    foundLicenseIdentifiers.add(field.replace(DEFINE_LICENSE, ""))
                } else if (field.startsWith(DEFINE_INT)) {
                    foundInternalLibraryIdentifiers.add(field.replace(DEFINE_INT, ""))
                } else if (field.startsWith(DEFINE_EXT)) {
                    foundExternalLibraryIdentifiers.add(field.replace(DEFINE_EXT, ""))
                }
            }
        }

        // add licenses
        // this has to happen first as the licenses need to be initialized before the libraries are read in
        for (licenseIdentifier in foundLicenseIdentifiers) {
            val license = genLicense(ctx, licenseIdentifier)
            if (license != null) {
                licenses.add(license)
            }
        }
        //add internal libs
        for (internalIdentifier in foundInternalLibraryIdentifiers) {
            val library = genLibrary(ctx, internalIdentifier)
            if (library != null) {
                library.isInternal = true
                internLibraries.add(library)
            }
        }

        //add external libs
        for (externalIdentifier in foundExternalLibraryIdentifiers) {
            val library = genLibrary(ctx, externalIdentifier)
            if (library != null) {
                library.isInternal = false
                externLibraries.add(library)
            }
        }
    }

    /**
     * This will summarize all libraries and elimate duplicates
     *
     * @param internalLibraries    the String[] with the internalLibraries (if set manual)
     * @param excludeLibraries     the String[] with the libs to be excluded
     * @param autoDetect           defines if the libraries should be resolved by their classpath (if possible)
     * @param checkCachedDetection defines if we should check the cached autodetected libraries (per version) (default: enabled)
     * @param sort                 defines if the array should be sorted
     * @return the summarized list of included Libraries
     */
    fun prepareLibraries(ctx: Context, internalLibraries: Array<out String>, excludeLibraries: Array<out String>, autoDetect: Boolean, checkCachedDetection: Boolean, sort: Boolean): ArrayList<Library> {
        val isExcluding = excludeLibraries.isNotEmpty()
        val libraries = HashMap<String, Library>()
        val resultLibraries = ArrayList<Library>()

        if (autoDetect) {
            val autoDetected = getAutoDetectedLibraries(ctx, checkCachedDetection)
            resultLibraries.addAll(autoDetected)

            if (isExcluding) {
                for (lib in autoDetected) {
                    libraries[lib.definedName] = lib
                }
            }
        }

        //Add all external libraries
        val extern = getExternLibraries()
        resultLibraries.addAll(extern)

        if (isExcluding) {
            for (lib in extern) {
                libraries[lib.definedName] = lib
            }
        }

        //Now add all libs which do not contains the info file, but are in the AboutLibraries lib
        if (internalLibraries != null) {
            for (internalLibrary in internalLibraries) {
                val lib = getLibrary(internalLibrary)
                if (lib != null) {
                    resultLibraries.add(lib)
                    libraries[lib.definedName] = lib
                }
            }
        }

        //remove libraries which should be excluded
        if (isExcluding) {
            for (excludeLibrary in excludeLibraries) {
                val lib = libraries[excludeLibrary]
                if (lib != null) {
                    resultLibraries.remove(lib)
                }
            }
        }

        if (sort) {
            Collections.sort(resultLibraries)
        }
        return resultLibraries
    }

    /**
     * Get all autoDetected Libraries
     *
     * @param ctx                  the current context
     * @param checkCachedDetection defines if we should check the cached autodetected libraries (per version) (default: enabled)
     * @return an ArrayList Library with all found libs by their classpath
     */
    fun getAutoDetectedLibraries(ctx: Context, checkCachedDetection: Boolean): List<Library> {
        val pi = ctx.getPackageInfo()
        val sharedPreferences = ctx.getSharedPreferences("aboutLibraries", Context.MODE_PRIVATE)
        val lastCacheVersion = sharedPreferences.getInt("versionCode", -1)
        val isCacheUpToDate = pi != null && lastCacheVersion == pi.versionCode

        if (checkCachedDetection) {//Retrieve from cache if up to date
            if (pi != null && isCacheUpToDate) {
                val autoDetectedLibraries = sharedPreferences.getString("autoDetectedLibraries", "")?.split(DELIMITER.toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray()

                if (autoDetectedLibraries?.isNotEmpty() == true) {
                    val libraries = ArrayList<Library>(autoDetectedLibraries.size)
                    for (autoDetectedLibrary in autoDetectedLibraries) {
                        val lib = getLibrary(autoDetectedLibrary)
                        if (lib != null) libraries.add(lib)
                    }
                    return libraries
                }
            }
        }

        val libraries = Detect.detect(ctx, libraries)
        if (pi != null && !isCacheUpToDate) {//Update cache
            val autoDetectedLibrariesPref = StringBuilder()

            for (lib in libraries) {
                autoDetectedLibrariesPref.append(DELIMITER).append(lib.definedName)
            }

            sharedPreferences.edit()
                    .putInt("versionCode", pi.versionCode)
                    .putString("autoDetectedLibraries", autoDetectedLibrariesPref.toString())
                    .apply()
        }

        return libraries
    }

    /**
     * Get all intern available Libraries
     *
     * @return an ArrayList Library with all available internLibraries
     */
    fun getInternLibraries(): ArrayList<Library> {
        return ArrayList(internLibraries)
    }

    /**
     * Get all extern available Libraries
     *
     * @return an ArrayList Library  with all available externLibraries
     */
    fun getExternLibraries(): ArrayList<Library> {
        return ArrayList(externLibraries)
    }

    /**
     * Get all available licenses
     *
     * @return an ArrayLIst License  with all available Licenses
     */
    fun getLicenses(): ArrayList<License> {
        return ArrayList(licenses)
    }

    /**
     * Get a library by its name (the name must be equal)
     *
     * @param libraryName the name of the lib (NOT case sensitiv) or the real name of the lib (this is the name used for github)
     * @return the found library or null
     */
    fun getLibrary(libraryName: String): Library? {
        for (library in libraries) {
            if (library.libraryName.toLowerCase() == libraryName.toLowerCase()) {
                return library
            } else if (library.definedName.toLowerCase() == libraryName.toLowerCase()) {
                return library
            }
        }
        return null
    }

    /**
     * Find a library by a searchTerm (Limit the results if there are more than one)
     *
     * @param searchTerm the term which is in the libs name (NOT case sensitiv) or the real name of the lib (this is the name used for github)
     * @param limit      -1 for all results or smaller 0 for a limitted result
     * @return an ArrayList Library with the found internLibraries
     */
    fun findLibrary(searchTerm: String, limit: Int): ArrayList<Library> {
        return find(libraries, searchTerm, false, limit)
    }

    /**
     * @param searchTerm
     * @param idOnly
     * @param limit
     * @return
     */
    fun findInInternalLibrary(searchTerm: String, idOnly: Boolean, limit: Int): ArrayList<Library> {
        return find(getInternLibraries(), searchTerm, idOnly, limit)
    }

    /**
     * @param searchTerm
     * @param idOnly
     * @param limit
     * @return
     */
    fun findInExternalLibrary(searchTerm: String, idOnly: Boolean, limit: Int): ArrayList<Library> {
        return find(getExternLibraries(), searchTerm, idOnly, limit)
    }

    /**
     * @param libraries
     * @param searchTerm
     * @param idOnly
     * @param limit
     * @return
     */
    private fun find(libraries: ArrayList<Library>, searchTerm: String, idOnly: Boolean, limit: Int): ArrayList<Library> {
        val localLibs = ArrayList<Library>()

        var count = 0
        for (library in libraries) {
            if (idOnly) {
                if (library.definedName.toLowerCase().contains(searchTerm.toLowerCase())) {
                    localLibs.add(library)
                    count = count + 1

                    if (limit != -1 && limit < count) {
                        break
                    }
                }
            } else {
                if (library.libraryName.toLowerCase().contains(searchTerm.toLowerCase()) || library.definedName.toLowerCase().contains(searchTerm.toLowerCase())) {
                    localLibs.add(library)
                    count = count + 1

                    if (limit != -1 && limit < count) {
                        break
                    }
                }
            }
        }

        return localLibs
    }


    /**
     * @param licenseName
     * @return
     */
    fun getLicense(licenseName: String): License? {
        for (license in getLicenses()) {
            if (license.licenseName.toLowerCase() == licenseName.toLowerCase()) {
                return license
            } else if (license.definedName.toLowerCase() == licenseName.toLowerCase()) {
                return license
            }
        }
        return null
    }

    /**
     * @param licenseName
     * @return
     */
    private fun genLicense(ctx: Context, licenseName: String): License? {
        var licenseName = licenseName
        licenseName = licenseName.replace("-", "_")

        return try {
            License(
                    licenseName,
                    getStringResourceByName(ctx, "license_" + licenseName + "_licenseName"),
                    getStringResourceByName(ctx, "license_" + licenseName + "_licenseWebsite"),
                    getStringResourceByName(ctx, "license_" + licenseName + "_licenseShortDescription"),
                    getStringResourceByName(ctx, "license_" + licenseName + "_licenseDescription")
            )
        } catch (ex: Exception) {
            Log.e("aboutlibraries", "Failed to generateLicense from file: " + ex.toString())
            null
        }

    }

    /**
     * @param libraryName
     * @return
     */
    private fun genLibrary(ctx: Context, libraryName: String): Library? {
        var libraryName = libraryName
        libraryName = libraryName.replace("-", "_")

        try {
            val lib = Library(definedName = libraryName, libraryName = getStringResourceByName(ctx, "library_" + libraryName + "_libraryName"))

            //Get custom vars to insert into defined areas
            val customVariables = getCustomVariables(ctx, libraryName)

            lib.author = getStringResourceByName(ctx, "library_" + libraryName + "_author")
            lib.authorWebsite = getStringResourceByName(ctx, "library_" + libraryName + "_authorWebsite")
            lib.libraryDescription = insertVariables(getStringResourceByName(ctx, "library_" + libraryName + "_libraryDescription"), customVariables)
            lib.libraryVersion = getStringResourceByName(ctx, "library_" + libraryName + "_libraryVersion")
            lib.libraryWebsite = getStringResourceByName(ctx, "library_" + libraryName + "_libraryWebsite")

            val licenseId = getStringResourceByName(ctx, "library_" + libraryName + "_licenseId")
            if (TextUtils.isEmpty(licenseId)) {
                val license = License("",
                        getStringResourceByName(ctx, "library_" + libraryName + "_licenseVersion"),
                        getStringResourceByName(ctx, "library_" + libraryName + "_licenseLink"),
                        insertVariables(getStringResourceByName(ctx, "library_" + libraryName + "_licenseContent"), customVariables),
                        insertVariables(getStringResourceByName(ctx, "library_" + libraryName + "_licenseContent"), customVariables)
                )
                lib.license = license
            } else {
                var license = getLicense(licenseId)
                if (license != null) {
                    license = license.copy()
                    license.licenseShortDescription = insertVariables(license.licenseShortDescription, customVariables)
                    license.licenseDescription = insertVariables(license.licenseDescription, customVariables)
                    lib.license = license
                }
            }

            lib.isOpenSource = java.lang.Boolean.valueOf(getStringResourceByName(ctx, "library_" + libraryName + "_isOpenSource"))
            lib.repositoryLink = getStringResourceByName(ctx, "library_" + libraryName + "_repositoryLink")

            lib.classPath = getStringResourceByName(ctx, "library_" + libraryName + "_classPath")

            return if (TextUtils.isEmpty(lib.libraryName) && TextUtils.isEmpty(lib.libraryDescription)) {
                null
            } else lib

        } catch (ex: Exception) {
            Log.e("aboutlibraries", "Failed to generateLibrary from file: " + ex.toString())
            return null
        }

    }

    /**
     * @param libraryName
     * @return
     */
    fun getCustomVariables(ctx: Context, libraryName: String): HashMap<String, String> {
        val customVariables = HashMap<String, String>()

        var customVariablesString = getStringResourceByName(ctx, DEFINE_EXT + libraryName)
        if (TextUtils.isEmpty(customVariablesString)) {
            customVariablesString = getStringResourceByName(ctx, DEFINE_INT + libraryName)
        }

        if (!TextUtils.isEmpty(customVariablesString)) {
            val customVariableArray = customVariablesString.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (customVariableArray.isNotEmpty()) {
                for (customVariableKey in customVariableArray) {
                    val customVariableContent = getStringResourceByName(ctx, "library_" + libraryName + "_" + customVariableKey)
                    if (!TextUtils.isEmpty(customVariableContent)) {
                        customVariables[customVariableKey] = customVariableContent
                    }
                }
            }
        }

        return customVariables
    }

    fun insertVariables(insertInto: String, variables: HashMap<String, String>): String {
        var insertInto = insertInto
        for ((key, value) in variables) {
            if (!TextUtils.isEmpty(value)) {
                insertInto = insertInto.replace("<<<" + key.toUpperCase() + ">>>", value)
            }
        }

        //remove the placeholder chars so the license is shown correct
        insertInto = insertInto.replace("<<<", "")
        insertInto = insertInto.replace(">>>", "")

        return insertInto
    }

    fun getStringResourceByName(ctx: Context, aString: String): String {
        val packageName = ctx.packageName

        val resId = ctx.resources.getIdentifier(aString, "string", packageName)
        return if (resId == 0) {
            ""
        } else {
            ctx.getString(resId)
        }
    }


    /**
     * @param modifications
     */
    fun modifyLibraries(modifications: HashMap<String, HashMap<String, String>>?) {
        if (modifications != null) {
            for ((key1, value1) in modifications) {
                var foundLibs: ArrayList<Library>? = findInExternalLibrary(key1, true, 1)
                if (foundLibs == null || foundLibs.size == 0) {
                    foundLibs = findInInternalLibrary(key1, true, 1)
                }

                if (foundLibs != null && foundLibs.size == 1) {
                    val lib = foundLibs[0]
                    for ((key2, value) in value1) {
                        val key = key2.toUpperCase()

                        if (key == LibraryFields.AUTHOR_NAME.name) {
                            lib.author = value
                        } else if (key == LibraryFields.AUTHOR_WEBSITE.name) {
                            lib.authorWebsite = value
                        } else if (key == LibraryFields.LIBRARY_NAME.name) {
                            lib.libraryName = value
                        } else if (key == LibraryFields.LIBRARY_DESCRIPTION.name) {
                            lib.libraryDescription = value
                        } else if (key == LibraryFields.LIBRARY_VERSION.name) {
                            lib.libraryVersion = value
                        } else if (key == LibraryFields.LIBRARY_WEBSITE.name) {
                            lib.libraryWebsite = value
                        } else if (key == LibraryFields.LIBRARY_OPEN_SOURCE.name) {
                            lib.isOpenSource = java.lang.Boolean.parseBoolean(value)
                        } else if (key == LibraryFields.LIBRARY_REPOSITORY_LINK.name) {
                            lib.repositoryLink = value
                        } else if (key == LibraryFields.LIBRARY_CLASSPATH.name) {
                            //note this can be set but won't probably work for autodetect
                            lib.classPath = value
                        } else if (key == LibraryFields.LICENSE_NAME.name) {
                            if (lib.license == null) {
                                lib.license = License("", "", "", "", "")
                            }
                            lib.license?.licenseName = value
                        } else if (key == LibraryFields.LICENSE_SHORT_DESCRIPTION.name) {
                            if (lib.license == null) {
                                lib.license = License("", "", "", "", "")
                            }
                            lib.license?.licenseShortDescription = value
                        } else if (key == LibraryFields.LICENSE_DESCRIPTION.name) {
                            if (lib.license == null) {
                                lib.license = License("", "", "", "", "")
                            }
                            lib.license?.licenseDescription = value
                        } else if (key == LibraryFields.LICENSE_WEBSITE.name) {
                            if (lib.license == null) {
                                lib.license = License("", "", "", "", "")
                            }
                            lib.license?.licenseWebsite = value
                        }
                    }
                }
            }
        }
    }

    companion object {

        val BUNDLE_THEME = "ABOUT_LIBRARIES_THEME"
        val BUNDLE_TITLE = "ABOUT_LIBRARIES_TITLE"
        val BUNDLE_STYLE = "ABOUT_LIBRARIES_STYLE"
        val BUNDLE_COLORS = "ABOUT_COLOR"

        private val DEFINE_LICENSE = "define_license_"
        private val DEFINE_INT = "define_int_"
        internal val DEFINE_EXT = "define_"

        private val DELIMITER = ";"
    }
}